examples/mfa/webauthn/server.go

```diff
@@ -0,0 +1,204 @@
+package webauthn
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/go-webauthn/webauthn/webauthn"
+	"github.com/openpubkey/openpubkey/client/cosigner/mfa"
+	"github.com/openpubkey/openpubkey/pktoken"
+	"github.com/openpubkey/openpubkey/util"
+)
+
+type Server struct {
+	uri      string
+	doneChan chan error
+	user     User
+	auth     *webauthn.WebAuthn
+	session  *webauthn.SessionData
+}
+
+type User interface {
+	webauthn.User
+	AddCredential(cred webauthn.Credential)
+}
+
+var _ mfa.Authenticator = (*Server)(nil)
+
+func New() (*Server, error) {
+	server := &Server{
+		doneChan: make(chan error),
+	}
+
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, fmt.Errorf("failed to bind to an available port: %w", err)
+	}
+	server.uri = fmt.Sprintf("http://localhost:%d", listener.Addr().(*net.TCPAddr).Port)
+
+	// WebAuthn configuration
+	cfg := &webauthn.Config{
+		RPDisplayName: "OpenPubkey",
+		RPID:          "localhost",
+		RPOrigin:      server.uri,
+	}
+
+	server.auth, err = webauthn.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	http.Handle("/", http.FileServer(http.Dir("webauthn/static")))
+
+	http.HandleFunc("/check-registration", server.checkIfRegistered)
+
+	http.HandleFunc("/register/begin", server.beginRegistration)
+	http.HandleFunc("/register/finish", server.finishRegistration)
+
+	http.HandleFunc("/login/begin", server.beginLogin)
+	http.HandleFunc("/login/finish", server.finishLogin)
+
+	http.HandleFunc("/done", server.done)
+
+	go func() {
+		http.Serve(listener, nil)
+	}()
+
+	return server, nil
+}
+
+func (s *Server) URI() string {
+	return s.uri
+}
+
+func (s *Server) Authenticate(pkt *pktoken.PKToken) error {
```

Comment:
IIRC we discussed that since we are still thinking through OpenPubkey POP Auth, that the example would not perform OpenPubkey authentication but only MFA authentication. I'm good with this.

To ensure we don't lose track of this, can you create a comment in the code and a github issue explaining that the MFA Auth example currently just naively trusts the identity in the PK Token. Then when we add OpenPubkey POP Auth to the MFA cosigner we can remove that comment and close that issue.

Reply:
cd30ca8